### DISCUSSION TITLE
Callback handling , How to use callback | Ref : https://github.com/spring-projects/spring-amqp/issues/1362

### DISCUSSION THREAD

**Discussion by kuldeepkalassts** [spring-projects/spring-amqp#1380] [Q&A] [answered]
Hi,

This is continuation of issue : https://github.com/spring-projects/spring-amqp/issues/1362 

1. How can we use and add callback when we get exception on secondary transaction ? from the code change we found that you now providing two method, one for enabling it and one for finding out if any exception occurred in commit.
2. How it is determined that which transaction is primary and which is secondary ?

for example : 
```java
@transactional
method foo(){
    sendRabbit();
    saveDB();
    sendRabbit();
    saveDB();
} 
```
SaveDB save data in DB and SendRabbit use template.convertAndSend in transactional enabled flow.

So which operation will be in  primary transaction

3. does similar "best effort 1PC" support in Spring's messaging that is available for RabbitMQ also available for following messaging technologies (as we are planning to migrate from RabbitMQ to one of these)
- Apache Kafka
- Google Pub Sub

**Comment by garyrussell** [comment, marked as answer]
The primary transaction is from the transaction manager associated with the `@Transactional` annotation.

See the documentation for how to use the callback https://docs.spring.io/spring-amqp/docs/current/reference/html/#tx-sync

Apache Kafka works the same way (although it does not currently have the callback we added here).

I can't answer for GPS; we don't maintain that project.
